data distribution editorial review (#4979)
Particular/docs.particular.net: nservicebus/messaging/data-distribution.md

---
title: Data distribution
summary: Why asynchronous messaging is not a good fit for data distribution scenarios
reviewed: 2020-09-01
redirects:
- nservicebus/azure/data-distribution
- samples/routing/data-distribution
---

Data distribution refers to a pattern where asynchronous messages (e.g. using NServiceBus) are used to deliver data to multiple targets in some coordinated fashion.

> A common example of a data distribution scenario is having cached data on multiple, scaled-out web servers, and attempting to deliver a message to each of them which indicates that they should drop their current cache entries and retrieve fresh data from the database.

Reviewer: I think comma placement here is an example of the putting-a-comma-where-you-would-pause-when-reading sin, not to actually separate fragments. It's long anyway - how about 2 sentences.
```suggestion
A common example of a data distribution scenario is having cached data on multiple scaled-out web servers and attempting to deliver a message to each of them. The message indicates each server should drop their current cache entries and retrieve fresh data from the database.
```

Author: > I think comma placement here is an example of the putting-a-comma-where-you-would-pause-when-reading sin, not to actually separate fragments.

I thought that the two adjectives should have commas in-between, but it turns out that's only the case when they are of [equal rank](https://www.grammarflip.com/curriculum/commas-between-two-adjectives), which these are not.

> It's long anyway - how about 2 sentences.

👍

## Timing and reliability

The goal of data distribution is generally to cause a change in data and/or behavior in multiple locations at the same time. NServiceBus does not provide any facility for receiving or processing messages in multiple locations at the same time. There is no urgency to act _now_ in reaction to a message sent with NServiceBus. Asynchronous messages are intended for eventual consistency. Each receiver processes a given message whenever it happens to receive that message from the queue.

In the worst-case scenario, some receivers of a data distribution message may succeed in handling the message, and others may fail, causing the message to be moved to the error queue. Until an administrator diagnoses the failure and returns the message to its source queue, transactions may be handled very differently by nodes that had successfully handled the message and those that had not.

These kinds of failures are difficult to track down. They depend upon more than the contents of a message and the code handling it. The results are also dependent upon the state of the data being handled by data distribution, which will not be consistent across all servers in a cluster.

> ## Differences to publish/subscribe

Reviewer: I can't back this up, but "difference to" / "different to" sounds unnatural to my ear. (Next paragraph and elsewhere too.) Is this American English vs. English English at play?

Author: > Is this American English vs. English English at play?

I think it [could be](https://english.stackexchange.com/a/845). I'll revert it to _from_.
